src/Microsoft.ML.TimeSeries/STL/Loess.cs

```diff
@@ -0,0 +1,206 @@
+                if (_x[mid] > xValue)
+                {
+                    end = mid;
+                }
+                else
+                {
+                    start = mid;
+                }
+            }
+            double distanceDiff = (_x[end] - xValue) - (xValue - _x[start]);
+
+            int index = distanceDiff > -NumericalThreshold ? start : end;
+            return _neighbors[index].Y(xValue);
+        }
+    }
+
+    /// <summary>
+    /// this class is used to define a set of weight functions. these functions are useful for various purposes for smoothing.
+    /// i.e., the weighted least squares.
+    /// </summary>
+    public class WeightMethod
+    {
+        /// <summary>
+        /// this is used for robust weight, it is one iteration step of loess.
+        /// </summary>
+        public static double BisquareWeight(double value)
+        {
+            double abs = Math.Abs(value);
+            if (abs >= 1)
+                return 0;
+            double temp = 1 - abs * abs;
+            return temp * temp;
+        }
+
+        /// <summary>
+        /// a famous weight function, since it enhances a chi-squared distributional approximation of an estimated of the error variance.
+        /// tricube should provide an adequate smooth in almost all situations. /ref
+        /// </summary>
+        public static double Tricube(double value)
+        {
+            double abs = Math.Abs(value);
+            if (abs >= 1)
+                return 0;
+            double temp = 1 - abs * abs * abs;
+            return temp * temp * temp;
+        }
+    }
+
+    /// <summary>
+    /// this class is used to store the parameters which are needed for lowess algorithm.
+    /// the name of these constansts are compliant with the original terms in paper.
+    /// </summary>
+    public class LoessConfiguration
+    {
+        /// <summary>
+        /// this value is used for performance concern. when the length of the series goes large, a ratio of neighbors will be significant,
+        /// which leads to unsatisfied slow. so this value is used to bound the maximum # of neighbors one epoch can have.
+        /// </summary>
+        public const int MaximumNeighborCount = 100;
+
+        /// <summary>
+        /// minumum number of neighbor counts, to apply underlying regression analysis.
+        /// this number should be even, so that neighbors on left/right side of a given data point is balanced. unbalanced neighbors would make the local-weighted regression biased noticeably at corner cases.
+        /// </summary>
+        public const int MinimumNeighborCount = 4;
+
+        /// <summary>
+        /// (0, 1], a smooth range ratio. let fn be the number of neighbors of a specific point.
+        /// </summary>
+        public static readonly double F = 0.3;
+
+        /// <summary>
+        /// the number of iterations for robust regression.
+        /// </summary>
+        public static readonly int T = 2;
```

Comment:
Just curious: is there a reason to make these 2 `static readonly` and the other 2 simply `const` ?

Reply:
Thanks for pointing out this. No special purpose, updated them to const.

---
In reply to: [440508775](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440508775) [](ancestors = 440508775)